5.0 updates to System.Text.Json docs (#21108)
dotnet/docs: docs/standard/serialization/system-text-json-how-to.md

::: zone-end

::: zone pivot="dotnet-core-3-1"

* By default, all public properties are serialized. You can [specify properties to exclude](#exclude-properties-from-serialization).
* The [default encoder](xref:System.Text.Encodings.Web.JavaScriptEncoder.Default) escapes non-ASCII characters, HTML-sensitive characters within the ASCII-range, and characters that must be escaped according to [the RFC 8259 JSON spec](https://tools.ietf.org/html/rfc8259#section-7).
* By default, JSON is minified. You can [pretty-print the JSON](#serialize-to-formatted-json).
* By default, casing of JSON names matches the .NET names. You can [customize JSON name casing](#customize-json-names-and-values).
* Circular references are detected and exceptions thrown.
* [Fields](../../csharp/programming-guide/classes-and-structs/fields.md) are excluded.
::: zone-end

Supported types include:
::: zone pivot="dotnet-5-0"

* .NET primitives that map to JavaScript primitives, such as numeric types, strings, and Boolean.
* User-defined [plain old CLR objects (POCOs)](https://en.wikipedia.org/wiki/Plain_old_CLR_object).
* One-dimensional and jagged arrays (`ArrayName[][]`).

> * `Dictionary<TKey,TValue>` where `TValue` is `object`, `JsonElement`, or a POCO.

Reviewer: We support more dictionary types than this.

```suggestion
```

> * Collections from the following namespaces.

Reviewer: ```suggestion
* Collections from the following namespaces, including dictionaries.
```

  * <xref:System.Collections>
  * <xref:System.Collections.Generic>

>   * <xref:System.Collections.Immutable>

Reviewer: In both 5.0 and 3.1 `S.C.Concurrent`, `.Specialized`, and `.ObjectModel` collections are also supported - https://github.com/dotnet/docs/issues/15807#issuecomment-576941811.

The following behaviors apply when deserializing JSON:

::: zone pivot="dotnet-5-0"

* By default, property name matching is case-sensitive. You can [specify case-insensitivity](#case-insensitive-property-matching).
* If the JSON contains a value for a read-only property, the value is ignored and no exception is thrown.

> * Non-public constructors are ignored by the serializer. However, parameterized constructors can be used if a parameterless constructor isn't available.

Reviewer: The point about parameterized ctors doesn't necessarily follow.

```suggestion
* Non-public constructors are ignored by the serializer. 
* A parameterized constructor can be used if a parameterless constructor isn't available and there is only one public parameterized constructor on the type. If `[JsonConstructorAttribute]` is used, the specified constructor will be used for deserialization regardless of the presence of other constructors.
```

* Deserialization to immutable objects or read-only properties is supported. See [Immutable types and Records](#immutable-types-and-records).
* By default, enums are supported as numbers. You can [serialize enum names as strings](#enums-as-strings).
* By default, fields are excluded. You can [include fields](#include-fields).
* By default, comments or trailing commas in the JSON throw exceptions. You can [allow comments and trailing commas](#allow-comments-and-trailing-commas).
* The [default maximum depth](xref:System.Text.Json.JsonReaderOptions.MaxDepth) is 64.

>     When you use System.Text.Json indirectly in an ASP.NET Core app, some default behaviors are different. For more information, see [Web defaults for JsonSerializerOptions](#web-defaults-for-jsonserializeroptions).

Reviewer: ```suggestion
When you use System.Text.Json indirectly in an ASP.NET Core app, some default behaviors are different. For more information, see [Web defaults for JsonSerializerOptions](#web-defaults-for-jsonserializeroptions).
```

>     When you use System.Text.Json indirectly in an ASP.NET Core app, some default behaviors are different. For more information, see [Web defaults for JsonSerializerOptions](#web-defaults-for-jsonserializeroptions).

Reviewer: This statement is also true for 3.1. Maybe add a corresponding note and inline details about case sensitivity and property naming policy?

By default, all public properties are serialized. If you don't want some of them to appear in the JSON output, you have several options. This section explains how to exclude:

::: zone pivot="dotnet-5-0"

* [Individual properties](#exclude-individual-properties)

> * [All read-only properties](#exclude-all-read-only-properties)

Author: The text that introduces the list specifies serialization, and the other links in the list also apply when serializing. Is there some reason to call it out specifically for this item in the list?

> * [All read-only properties](#exclude-all-read-only-properties)

Reviewer: Per your comment and after re-reading the section, I withdraw the suggestion.

}
```

::: zone pivot="dotnet-5-0"
You can specify conditional exclusion by setting the [[JsonIgnore]](xref:System.Text.Json.Serialization.JsonIgnoreAttribute) attribute's `Condition` property. The <xref:System.Text.Json.Serialization.JsonIgnoreCondition> enum provides the following options:

* `Always` - The property will always be ignored. If no `Condition` is specified, this option is assumed.

> * `Never` - The property will always be serialized and deserialized, regardless of the `IgnoreNullValues` global setting.

Reviewer: ```suggestion
* `Never` - The property will always be serialized and deserialized, regardless of the `DefaultIgnoreCondition` global setting.
```

This option applies only to serialization. During deserialization, read-only properties are ignored by default.

### Exclude all null-value properties

::: zone pivot="dotnet-5-0"
To exclude all null-value properties, set the <xref:System.Text.Json.JsonSerializerOptions.DefaultIgnoreCondition> property to <xref:System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull>, as shown in the following example:

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/IgnoreNullOnSerialize.cs" highlight="28":::

> This setting applies to serialization. For information about deserialization, see [Ignore null when deserializing](#ignore-null-when-deserializing).

Author: I'll remove the section.

}
Date: 10/19/2020 2:40:27 PM
TemperatureC: 40
Summary: Hot
```

When you use `System.Text.Json` indirectly through ASP.NET Core, quoted numbers are allowed because ASP.NET Core specifies [web default options](xref:System.Text.Json.JsonSerializerDefaults.Web).

::: zone-end

::: zone pivot="dotnet-core-3-1"
<xref:System.Text.Json> 3.1 doesn' support serializing or deserializing numbers in quotes.
::: zone-end

## Copy a JsonSerializerOptions instance

::: zone pivot="dotnet-5-0"
There is a <xref:System.Text.Json.JsonSerializerOptions> constructor that lets you create a new instance with the same options as an existing instance, as shown in the following example:

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/CopyOptions.cs":::

The preceding code produces output similar to the following example:

```output
Output JSON:
{
  "Date": "2020-10-20T18:38:33.4153344-07:00",
  "TemperatureC": 40,
  "Summary": "Hot"
}
```

::: zone-end

::: zone pivot="dotnet-core-3-1"
A constructor that takes an existing instance is not available in `System.Text.Json` 3.1.
::: zone-end

## Web defaults for JsonSerializerOptions

::: zone pivot="dotnet-5-0"

> There is a <xref:System.Text.Json.JsonSerializerOptions> constructor that creates a new instance with [the default options that are used for web apps](xref:System.Text.Json.JsonSerializerDefaults.Web), as shown in the following example:

Reviewer 2: Technically the constructor accepts a JsonSerializerDefaults object. It could be Web or something else, depending on what the user specifies.

Author: Good point, I changed the wording so it doesn't imply this is the only possible usage of this constructor, though as a practical matter web defaults is the only scenario it's useful for at present.

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/OptionsDefaults.cs":::

The preceding code produces output similar to the following example:

```output
options.PropertyNameCaseInsensitive: True
options.JsonNamingPolicy: System.Text.Json.JsonCamelCaseNamingPolicy
options.NumberHandling: AllowReadingFromString
Output JSON:
{
  "date": "2020-10-20T18:38:33.4211057-07:00",
  "temperatureC": 40,
  "summary": "Hot"
}
Date: 10/20/2020 6:38:33 PM
TemperatureC: 40
Summary: Hot
```

::: zone-end

::: zone pivot="dotnet-core-3-1"

> A constructors that specifies web defaults is not available in `System.Text.Json` 3.1.

Reviewer 2: ```suggestion
A constructor that specifies web defaults is not available in `System.Text.Json` 3.1.
```

Null values in the JSON are ignored only if they are valid. Null values for non-nullable value types cause exceptions.

## Preserve references and handle circular references

::: zone pivot="dotnet-5-0"

To preserve references and handle circular references, set <xref:System.Text.Json.JsonSerializerOptions.ReferenceHandler%2A> to <xref:System.Text.Json.Serialization.ReferenceHandler.Preserve%2A>. This setting causes the following behavior:

* On serialize:

  When writing complex types, the serializer also writes metadata properties (`$id`, `$values`, and `$ref`).

* On deserialize:

  Metadata is expected (although not mandatory), and the deserializer tries to understand it.

The following code illustrates use of the `Preserve` setting.

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/PreserveReferences.cs" highlight="34":::

Here are some other notes about preserving references:

* `MaxDepth` validation is not affected.
* The following types can be serialized as references but can't be deserialized as references:
  * Immutable types
  * Arrays

>   * Value types, such as structs

Reviewer 3: > We should also mention that we throw when reference metadata is found when deserializing these types.

We throw if an Immutable type or Array contains $`ref` or `$id`.
POCO value types can ignore `$id` but will throw if they contain `$ref`.

>   * Value types, such as structs

Reviewer 3: @tdykstra I think it would say: "this feature cannot be used to preserve value types and/or immutable types"
Reasons: 
* Value types: we don't have a reference per se.
* Immutable types: they are created until the payload is fully read, which makes it impossible to preserve a reference of itself that could be within the payload e.g: 
```cs
var options = new JsonSerializerOptions { ReferenceHandler = ReferenceHandler.Preserve };
var obj = JsonSerializer.Deserialize<ImmutableDictionary<string, object>>(@"{""$id"": ""1"", ""property1"": {""$ref"": ""1""} }", options);
```

> * To determine if objects are equal, `System.Text.Json` uses the default equality of `Hashset` or `Dictionary`, which calls <xref:System.Object.GetHashCode%2A?displayProperty=nameWithType> or <xref:System.Object.Equals%2A?displayProperty=nameWithType>.

Reviewer 3: We don't use the same equality comparer as the Dictionary or Hashset, we use [`ReferenceEqualityComparer.Instance`](https://docs.microsoft.com/dotnet/api/system.collections.generic.referenceequalitycomparer.instance?view=net-5.0)

From docs:
> An IEqualityComparer<T> that uses reference equality (ReferenceEquals(Object, Object)) instead of value equality (Equals(Object)) when comparing two object instances.

> * An exception is thrown if there are properties other than `$ref` in a JSON reference object.

Reviewer 3: > What does "JSON reference object" mean?

We should probably say "JSON object that contains a $ref metadata" since that's what we use on the [`ReferenceHandler.Preserve`](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.referencehandler.preserve?view=net-5.0) notes.

Those other three bullets that you mentioned are already specified in said doc.

> * An exception is thrown if there are properties other than `$ref` in a JSON reference object.

Reviewer: Nice, we should link to this doc then: https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.referencehandler.preserve?view=net-5.0.

> * An exception is thrown if there are properties other than `$ref` in a JSON reference object.

Reviewer 3: We should at least link to [ReferenceHandler.Preserve](https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.referencehandler.preserve?view=net-5.0) in this doc.

::: zone-end

::: zone pivot="dotnet-core-3-1"
<xref:System.Text.Json> 3.1 doesn't support serializing or deserializing numbers in quotes.
::: zone-end

## Immutable types and Records

::: zone pivot="dotnet-5-0"
`System.Text.Json` can use a parameterized constructor, which makes it possible to deserialize an immutable class or struct. For a class, if the only constructor is a parameterized one, that constructor will be used. For a struct, or a class with multiple constructors, specify the one to use by applying the [[JsonConstructor]](xref:System.Text.Json.Serialization.JsonConstructorAttribute.%23ctor%2A) attribute, as shown in the following example:

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/ImmutableTypes.cs" highlight="13":::

Records in C# 9 are also supported, as shown in the following example:

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/Records.cs":::
::: zone-end

::: zone pivot="dotnet-core-3-1"
`JsonConstructorAttribute` and C# 9 Record support aren't available in `System.Text.Json` 3.1.
::: zone-end

## Copy JsonSerializerOptions

::: zone pivot="dotnet-5-0"

> There is a <xref:System.Text.Json.JsonSerializerOptions> constructor that lets you create a new instance with the same options as an existing instance, as shown in the following example:

Reviewer: Can we link to the docs for the ctor?

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/CopyOptions.cs" highlight="29":::
::: zone-end

::: zone pivot="dotnet-core-3-1"
A constructor that takes an existing instance is not available in `System.Text.Json` 3.1.
::: zone-end

## Web defaults for JsonSerializerOptions

::: zone pivot="dotnet-5-0"
Here are the options that have different defaults for web apps:

* <xref:System.Text.Json.JsonSerializerOptions.PropertyNameCaseInsensitive%2A> = `true`
* <xref:System.Text.Json.JsonNamingPolicy> = <xref:System.Text.Json.JsonNamingPolicy.CamelCase>
* <xref:System.Text.Json.JsonSerializerOptions.NumberHandling%2A> = <xref:System.Text.Json.Serialization.JsonNumberHandling.AllowReadingFromString>

> There's a <xref:System.Text.Json.JsonSerializerOptions> constructor that lets you create a new instance with [the default options that ASP.NET Core uses for web apps](xref:System.Text.Json.JsonSerializerDefaults.Web), as shown in the following example:

Reviewer: Perhaps we should generalize this section to be about the constructor, and the mention web defaults as one possilbe param. This should scale well if we have more defaults later on.

Author: > generalize this section to be about the constructor, and then mention web defaults as one possible param. This should scale well if we have more defaults later on.

The emphasis in the conceptual docs is covering scenarios rather than APIs -- for constructor documentation we have the API ref, and for the scenario of needing to apply web defaults we have this section of a conceptual doc, which explains what the defaults are and how to apply them. An alternative is  to just note what the defaults are (since that doesn't show up in the API ref) and link to the constructor API ref, and omit the code example.

Null values in the JSON are ignored only if they are valid. Null values for non-nullable value types cause exceptions.

## Preserve references and handle circular references

::: zone pivot="dotnet-5-0"

To preserve references and handle circular references, set <xref:System.Text.Json.JsonSerializerOptions.ReferenceHandler%2A> to <xref:System.Text.Json.Serialization.ReferenceHandler.Preserve%2A>. This setting causes the following behavior:

* On serialize:

  When writing complex types, the serializer also writes metadata properties (`$id`, `$values`, and `$ref`).

* On deserialize:

  Metadata is expected (although not mandatory), and the deserializer tries to understand it.

The following code illustrates use of the `Preserve` setting.

:::code language="csharp" source="snippets/system-text-json-how-to-5-0/csharp/PreserveReferences.cs" highlight="34":::

This feature can't be used to preserve value types or immutable types. On deserialization, the instance of an immutable type is created after the entire payload is read. So it would be impossible to deserialize the same instance if a reference to it appears within the JSON payload.

> For value types, immutable types, and arrays, no reference metadata is serialized. On deserialization, an exception is thrown if `$ref` or `$id` is found. However, POCO value types ignore `$id` to make it possible to deserialize payloads that were serialized by using Newtonsoft.Json.  Newtonsoft.Json does serialize metadata for such types.

Reviewer 3: We have other validations that interfere with the one described in the paragraph.

* For arrays: we validate first that `$id` is read before `$values`.
* For objects: we throw if we read `$values`.

This one is fine as it is, also IIRC value type collections with $id and $values can be deserialized correctly by ignoring the preserve semantics.

> For value types, immutable types, and arrays, no reference metadata is serialized. On deserialization, an exception is thrown if `$ref` or `$id` is found. However, POCO value types ignore `$id` to make it possible to deserialize payloads that were serialized by using Newtonsoft.Json.  Newtonsoft.Json does serialize metadata for such types.

Reviewer 3: ```suggestion
For value types, immutable types, and arrays, no reference metadata is serialized. On deserialization, an exception is thrown if `$ref` or `$id` is found. However, value types ignore `$id` (and `$values` in the case of collections) to make it possible to deserialize payloads that were serialized by using Newtonsoft.Json.  Newtonsoft.Json does serialize metadata for such types.
```

> For value types, immutable types, and arrays, no reference metadata is serialized. On deserialization, an exception is thrown if `$ref` or `$id` is found. However, POCO value types ignore `$id` to make it possible to deserialize payloads that were serialized by using Newtonsoft.Json.  Newtonsoft.Json does serialize metadata for such types.

Author: >This one is fine as it is

So the recommendation is to not mention $values here, or add remarks per the two bullet points about arrays and objects?

> For value types, immutable types, and arrays, no reference metadata is serialized. On deserialization, an exception is thrown if `$ref` or `$id` is found. However, POCO value types ignore `$id` to make it possible to deserialize payloads that were serialized by using Newtonsoft.Json.  Newtonsoft.Json does serialize metadata for such types.

Reviewer 3: I left a suggestion but wasn't posted as part of this thread:
https://github.com/dotnet/docs/pull/21108#discussion_r517637520

> To determine if objects are equal, `System.Text.Json` uses <xref:System.Collections.Generic.ReferenceEqualityComparer.Instance%2A>, which uses reference equality (<xref:System.Object.ReferenceEquals(System.Object,System.Object)>) instead of value equality (<xref:System.Object.Equals(System.Object)?displayProperty=nameWithType> when comparing two object instances.

Reviewer 3: @layomia this should render to what you are asking here:
`<xref:System.Collections.Generic.ReferenceEqualityComparer.Instance?displayProperty=nameWithType>`

For more information about how references are serialized and deserialized, see <xref:System.Text.Json.Serialization.ReferenceHandler.Preserve%2A?displayProperty=nameWithType>.

> The <xref:System.Text.Json.Serialization.ReferenceResolver> class defines the behavior of preserving references on serialization and deserialization. Create a derived class to specify custom behavior. For an example, see [GuidReferenceResolver](https://github.com/dotnet/dotnet-docs/blob/master/docs/standard/serialization/snippets/system-text-json-how-to-5-0/csharp/GuidReferenceResolver.cs).

Author: Yes, this will be in the dotnet/docs repo after the PR is merged and I will go back and update this link.
